control_panel_api/authentication.py

```diff
@@ -1,37 +1,69 @@
+import logging
+
+import requests
 from django.conf import settings
-import jwt
-from jwt.exceptions import InvalidTokenError
+from jose import jwt
+from jose.exceptions import JWTError
+from requests.exceptions import RequestException
 from rest_framework.authentication import (
     BaseAuthentication,
-    get_authorization_header
+    get_authorization_header,
 )
 from rest_framework.exceptions import AuthenticationFailed
 
 from control_panel_api.models import User
 
+logger = logging.getLogger(__name__)
 
-class Auth0JWTAuthentication(BaseAuthentication):
 
-    def authenticate(self, request):
+def get_jwks():
+    response = requests.get(settings.OIDC_WELL_KNOWN_URL)
+    return response.json()
+
+
+def get_key(jwks, unverified_header):
+    for jwk in jwks.get('keys'):
+        if jwk['kid'] == unverified_header['kid']:
+            return jwk
 
+    return None
+
+
+class Auth0JWTAuthentication(BaseAuthentication):
+    def authenticate(self, request):
         try:
             prefix, token = get_authorization_header(request).split()
-
         except (ValueError, TypeError):
             return None
 
         if prefix != 'JWT'.encode():
             return None
 
+        try:
+            unverified_header = jwt.get_unverified_header(token)
+        except JWTError as e:
+            raise AuthenticationFailed('Error decoding JWT header') from e
+
+        try:
+            jwks = get_jwks()
+        except RequestException as e:
+            logger.error(e)
+            raise AuthenticationFailed(e) from e
+
+        key = get_key(jwks, unverified_header)
+        if key is None:
+            raise AuthenticationFailed('kid matching failure')
+
         try:
```

Comment:
I'd move this logic and exception handling in a method, something called `get_decoded_jwt()` or something like that.

This would make this `authenticate()` method a little big smaller.

Reply:
Yeh I think in this case I can put `get_key` and `get_jwks` together. Although usually I'd prefer not to couple it.

Reply:
Rather than put the two existing functions together, which I think are already well factored, just add another function which calls them